src/protocol/ErrorFactory.ts

```diff
@@ -66,119 +76,119 @@ export class ClientErrorFactory {
-            (m, c) => new TargetNotMemberError(m, c));
+            (m, c, e) => new TargetNotMemberError(m, c, e));
         this.register(ClientProtocolErrorCodes.TOPIC_OVERLOAD,
-            (m, c) => new TopicOverloadError(m, c));
+            (m, c, e) => new TopicOverloadError(m, c, e));
         this.register(ClientProtocolErrorCodes.TRANSACTION,
-            (m, c) => new TransactionError(m, c));
+            (m, c, e) => new TransactionError(m, c, e));
         this.register(ClientProtocolErrorCodes.TRANSACTION_NOT_ACTIVE,
-            (m, c) => new TransactionNotActiveError(m, c));
+            (m, c, e) => new TransactionNotActiveError(m, c, e));
         this.register(ClientProtocolErrorCodes.TRANSACTION_TIMED_OUT,
-            (m, c) => new TransactionTimedOutError(m, c));
+            (m, c, e) => new TransactionTimedOutError(m, c, e));
         this.register(ClientProtocolErrorCodes.UNSUPPORTED_OPERATION,
-            (m, c) => new UnsupportedOperationError(m, c));
+            (m, c, e) => new UnsupportedOperationError(m, c, e));
         this.register(ClientProtocolErrorCodes.NO_DATA_MEMBER,
-            (m, c) => new NoDataMemberInClusterError(m, c));
+            (m, c, e) => new NoDataMemberInClusterError(m, c, e));
         this.register(ClientProtocolErrorCodes.STALE_TASK_ID,
-            (m, c) => new StaleTaskIdError(m, c));
+            (m, c, e) => new StaleTaskIdError(m, c, e));
         this.register(ClientProtocolErrorCodes.FLAKE_ID_NODE_ID_OUT_OF_RANGE_EXCEPTION,
-            (m, c) => new NodeIdOutOfRangeError(m, c));
+            (m, c, e) => new NodeIdOutOfRangeError(m, c, e));
         this.register(ClientProtocolErrorCodes.CONSISTENCY_LOST_EXCEPTION,
-            (m, c) => new ConsistencyLostError(m, c));
-        this.register(ClientProtocolErrorCodes.INDETERMINATE_OPERATION_STATE,
-            (m, c) => new IndeterminateOperationStateError(m, c));
+            (m, c, e) => new ConsistencyLostError(m, c, e));
     }
 
     createErrorFromClientMessage(clientMessage: ClientMessage): Error {
         const errorHolders = ErrorsCodec.decode(clientMessage);
         return this.createError(errorHolders, 0);
     }
 
-    createError(errorHolders: ErrorHolder[], errorHolderIdx: number): Error {
+    private createError(errorHolders: ErrorHolder[], errorHolderIdx: number): Error {
         if (errorHolderIdx === errorHolders.length) {
             return null;
         }
         const errorHolder = errorHolders[errorHolderIdx];
         const factoryFn = this.codeToErrorConstructor.get(errorHolder.errorCode);
         let error: Error;
         if (factoryFn != null) {
-            error = factoryFn(errorHolder.message, this.createError(errorHolders, errorHolderIdx + 1));
+            error = factoryFn(errorHolder.message, this.createError(errorHolders, errorHolderIdx + 1), errorHolder);
         } else {
             error = new UndefinedErrorCodeError(errorHolder.message, errorHolder.className);
```

Comment:
I see that once we are faced with an unknown error code, we stop iterating over the error holders. I know that it's a long shot but there may be other elements in the error tree that we can extract something. Maybe, we could continue iterating while setting the `cause` of `UndefinedErrorCodeError` to the next element. What do you think about it?

Reply:
I didn't touch the old logic, but that's a nice suggestion. Implemented it in b947edaf337744e1a0ab6b3976abd6ca5ade386f